src/ray/gcs/gcs_client/gcs_client.cc

```diff
@@ -260,90 +266,146 @@ bool GcsClient::GetGcsServerAddressFromRedis(redisContext *context,
+    // mechanism: if the address does not change (meaning gcs server doesn't restart),
+    // the connection can be made at most once in
+    // `minimum_gcs_reconnect_interval_milliseconds` milliseconds.
+    if (last_reconnect_address_ == address &&
+        (current_sys_time_ms() - last_reconnect_timestamp_ms_) <
+            RayConfig::instance().minimum_gcs_reconnect_interval_milliseconds()) {
+      RAY_LOG(DEBUG)
+          << "Repeated reconnection in "
+          << RayConfig::instance().minimum_gcs_reconnect_interval_milliseconds()
+          << " milliseconds, return directly.";
+      ExecuteReconnectedCallbacks();
+      reconnect_in_progress_ = false;
       return;
     }
-    if (get_server_address_func_(&address)) {
-      // After GCS is restarted, the gcs client will reestablish the connection. At
-      // present, every failed RPC request will trigger `ReconnectGcsServer`. In order to
-      // avoid repeated connections in a short period of time, we add a protection
-      // mechanism: if the address does not change (meaning gcs server doesn't restart),
-      // the connection can be made at most once in
-      // `minimum_gcs_reconnect_interval_milliseconds` milliseconds.
-      if (last_reconnect_address_ == address &&
-          (current_sys_time_ms() - last_reconnect_timestamp_ms_) <
-              RayConfig::instance().minimum_gcs_reconnect_interval_milliseconds()) {
-        RAY_LOG(DEBUG)
-            << "Repeated reconnection in "
-            << RayConfig::instance().minimum_gcs_reconnect_interval_milliseconds()
-            << " milliseconds, return directly.";
-        return;
-      }
-
-      RAY_LOG(DEBUG) << "Attemptting to reconnect to GCS server: " << address.first << ":"
-                     << address.second;
-      if (Ping(address.first, address.second, 100)) {
-        // If `last_reconnect_address_` port is -1, it means that this is the first
-        // connection and no log will be printed.
-        if (last_reconnect_address_.second != -1) {
-          RAY_LOG(INFO) << "Reconnected to GCS server: " << address.first << ":"
-                        << address.second;
-        }
-        reconnected = true;
-        break;
-      }
-    }
-    std::this_thread::sleep_for(
-        std::chrono::milliseconds(kGCSReconnectionRetryIntervalMs));
+
+    RAY_LOG(DEBUG) << "Attemptting to reconnect to GCS server: " << address.first << ":"
+                   << address.second;
+    gcs_rpc_client_->ResetHeartBeatInfoClient(address.first, address.second,
+                                              *client_call_manager_);
+    rpc::CheckAliveRequest request;
+    gcs_rpc_client_->CheckAlive(
+        request,
+        [this, start, address](const Status &status, const rpc::CheckAliveReply &reply) {
+          RAY_LOG(DEBUG) << "CheckAlive returns " << status;
+          OnReconnectionFinished(status, start, address);
+        });
+  } else {
+    RAY_LOG(WARNING) << "Can not get gcs server address.";
```

Comment:
Can we move this to up? LIke 

```
if (!get_server_address_func_(&address)) {
    ExecuteReconnectedCallbacks();
    reconnect_in_progress_ = false;
}

RAY_LOG(DEBUG) << "Attemptting to reconnect to GCS server: " << address.first << ":"
               << address.second;
gcs_rpc_client_->ResetHeartBeatInfoClient(address.first, address.second,
                                          *client_call_manager_);

```

Also, `ExecuteReconnectedCallbacks ` doesn't match the signature in API definition..